### PR TITLE
New PushNotificationDataSource

### PR DESCRIPTION
**Pull request by nelsitoPuglisi**
## Goal

Log push notifications as span events in the session span instead of breadcrumbs

## Testing

Integration and manual tests

### CODE REVIEW

**Comment by nelsitoPuglisi** [comment]
* **#688** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/688?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @nelsitoPuglisi and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: embrace-android-sdk/src/integrationTest/kotlin/io/embrace/android/embracesdk/testcases/PublicApiTest.kt [hunk 1]
```diff
@@ -4,9 +4,16 @@ import android.os.Build.VERSION_CODES.TIRAMISU
...
+import io.embrace.android.embracesdk.arch.schema.EmbType
+import io.embrace.android.embracesdk.findEventOfType
+import io.embrace.android.embracesdk.findSessionSpan
+import io.embrace.android.embracesdk.getSentSessionMessages
+import io.embrace.android.embracesdk.hasEventOfType
 import io.embrace.android.embracesdk.internal.ApkToolsConfig
 import io.embrace.android.embracesdk.internal.IdGeneratorTest.Companion.validPattern
+import io.embrace.android.embracesdk.internal.clock.millisToNanos
 import io.embrace.android.embracesdk.recordSession
+import org.junit.Assert
```
Comment (fractalwrench) [inline embrace-android-sdk/src/integrationTest/kotlin/io/embrace/android/embracesdk/testcases/PublicApiTest.kt:16]: nit: I'd assume these are all unused imports?

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/payload/Breadcrumbs.kt [hunk 2]
```diff
@@ -25,11 +25,5 @@ internal data class Breadcrumbs(
      * List of RN Action breadcrumbs.
      */
     @Json(name = "rna")
-    val rnActionBreadcrumbs: List<RnActionBreadcrumb>? = null,
-
-    /**
-     * List of captured push notifications
-     */
-    @Json(name = "pn")
-    val pushNotifications: List<PushNotificationBreadcrumb>? = null,
```
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/payload/Breadcrumbs.kt:34]: Can the `LegacyPushNotificationDataSource` be deleted if we're getting rid of payload fields in the same changeset?

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/injection/DataSourceModule.kt [hunk 3]
```diff
@@ -69,6 +71,19 @@ internal class DataSourceModuleImpl(
         )
     }
 
+    override val pushNotificationDataSource: DataSourceState<PushNotificationDataSource> by dataSourceState {
+        DataSourceState(
```
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/injection/DataSourceModule.kt:75]: How do users enable this feature in their apps? I'm a little surprised that there isn't any parameter supplied for `configGate` although if that's existing behavior that's ok

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/capture/crumbs/EmbraceBreadcrumbService.kt [hunk 4]
```diff
@@ -108,14 +105,18 @@ internal class EmbraceBreadcrumbService(
...
-    ) = pushNotificationBreadcrumbDataSource.logPushNotification(
-        title,
-        body,
-        topic,
-        id,
-        notificationPriority,
-        type
-    )
+    ) {
+        dataSourceModuleProvider()?.pushNotificationDataSource?.dataSource?.apply {
```
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/capture/crumbs/EmbraceBreadcrumbService.kt:109]: nit: you could remove the apply here as there's only one statement inside

File: embrace-android-sdk/src/integrationTest/kotlin/io/embrace/android/embracesdk/testcases/PushNotificationTest.kt [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        Assert.assertNotNull(event)
+        Assert.assertTrue(event.timestampNanos > 0)
+        Assert.assertEquals(
+            mapOf(
+                EmbType.System.PushNotification.toEmbraceKeyValuePair(),
+                "notification.title" to "",
+                "notification.type" to type,
+                "notification.body" to "",
+                "notification.id" to "id",
+                "notification.from" to "",
```
Comment (fractalwrench) [inline embrace-android-sdk/src/integrationTest/kotlin/io/embrace/android/embracesdk/testcases/PushNotificationTest.kt:103]: These empty string assertions don't look right - if `"title"` and `"body"` are being supplied as parameters in the test invocation then I'd expect them to show up in the attrs map.

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/capture/crumbs/PushNotificationDataSource.kt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                        id,
+                        notificationPriority,
+                        type.type,
+                        clock.now()
+                    )
+                    addEvent(crumb, ::toSpanEventData)
+                }
+            )
+        } catch (ex: Exception) {
+            logger.logError("Failed to log push notificaction $id", ex)
```
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/capture/crumbs/PushNotificationDataSource.kt:58]:
```suggestion
            logger.logError("Failed to log push notification $id", ex)
```

**Review by fractalwrench — COMMENTED** [review]
I had one query about some assertions in the test cases, otherwise this looks good

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/injection/DataSourceModule.kt (hunk 3 above)
Comment (nelsitoPuglisi) [reply embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/injection/DataSourceModule.kt:75]: This is used by manually calling our public api, and automacally if the feature is enabled with this guide: https://embrace.io/docs/android/features/push-notifications/

File: embrace-android-sdk/src/integrationTest/kotlin/io/embrace/android/embracesdk/testcases/PublicApiTest.kt (hunk 1 above)
Comment (nelsitoPuglisi) [reply embrace-android-sdk/src/integrationTest/kotlin/io/embrace/android/embracesdk/testcases/PublicApiTest.kt:16]: 🤦

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/injection/DataSourceModule.kt [hunk 7]
```diff
@@ -69,6 +71,20 @@ internal class DataSourceModuleImpl(
...
+            factory = {
+                System.out.println("DataSourceModuleImpl ${essentialServiceModule.configService}")
```
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/injection/DataSourceModule.kt:77]: Would suggest removing this `println` statement

**Review by fractalwrench — APPROVED** [review]
LGTM

**Pull request merged**